### PR TITLE
Cypress tests

### PR DESCRIPTION
**Pull request by suryapulakhandam**
## Description
added tests for A_Z nav tests, action-links, breadcrumbs, contents-list, details, do-dont-list, emergency-alert, footer, hero, images, inset-text, list-panel, pagination, panel, promo, review-date

### CODE REVIEW

File: tests/cypress/cypress.json [hunk 1]
```diff
@@ -0,0 +1,50 @@
+{
```
Comment (AdamChrimes) [inline tests/cypress/cypress.json]: Does this file need all these fields? Also the projectID seems a little random can this be more related to the project?
There also seems to be duplication of the iPhone viewport sizes and those are iPhone 5 sizes rather than iPhone4

Cleaner configuration file:

```
{
  "projectId": "nhsuk-frontend",
  "baseUrl": "http://localhost:3000",
  "chromeWebSecurity": false,
  "iPhone 5":
  {
     "viewportWidth": 320,
    "viewportHeight": 480
  },
  "baseUrl": "http://localhost:3000/components/",
  "liveUrl": "https://www.nhs.uk/",
}
```

File: tests/cypress/integration/components/a-z-nav.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
+describe('setting screen res', function () {
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/a-z-nav.js]: This should describe the whole test such as "Navigation A-Z tests" rather than setting the screen resolution, as this describes the whole tests not just the first action.

File: tests/cypress/integration/components/a-z-nav.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+  it('Navigating to A-Z Navigation', function () {
+
+    cy
+      .visit(Cypress.config().baseUrl + '/nav-a-z.html')
+
+  });
+
+
+  // checking for class name - nhsuk-c-a-z-nav__list
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/a-z-nav.js]: Do we need comments here? As it is described in plain english in the `it('Checking for class name - nhsuk-c-a-z-nav__list'`

File: tests/cypress/integration/components/a-z-nav.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  it('Checking for class name - nhsuk-c-a-z-nav__list', function () {
+    cy
+      .get('#nhsuk-nav-a-z').children().should('have.class', 'nhsuk-nav-a-z__list')
+
+  });
+
+  // checking for class name - nhsuk-c-a-z-nav__list-item-link
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/a-z-nav.js]: See above about comments.

File: tests/cypress/integration/components/a-z-nav.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  it('Checking for class name - nhsuk-nav-a-z__link', function () {
+    cy
+      .get('.nhsuk-nav-a-z__item').first().children().should('have.class', 'nhsuk-nav-a-z__link')
+
+  });
+
+  // check for aria label - A- Z Navigation
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/a-z-nav.js]: See above about comments.

File: tests/cypress/integration/components/a-z-nav.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  it('Checking Aria label - A- Z navigation', function () {
+    cy
+      .get('#nhsuk-nav-a-z').should('have.attr', 'aria-label', 'A-Z Navigation')
+
+  });
+
+
+  // checking for role - navigation
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/a-z-nav.js]: See above about comments.

File: tests/cypress/integration/components/a-z-nav.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
+describe('setting screen res', function () {
+
+  beforeEach(function () {
+    cy.viewport('macbook-15');
+
+  });
+
+  it('Navigating to A-Z Navigation', function () {
+
+    cy
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/a-z-nav.js]: Can this go on the same line `cy.visit()`?

File: tests/cypress/integration/components/a-z-nav.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
+describe('setting screen res', function () {
+
+  beforeEach(function () {
+    cy.viewport('macbook-15');
+
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/a-z-nav.js]: Extra empty new line

File: tests/cypress/integration/components/a-z-nav.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  });
+
+  it('Navigating to A-Z Navigation', function () {
+
+    cy
+      .visit(Cypress.config().baseUrl + '/nav-a-z.html')
+
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/a-z-nav.js]: Extra empty new line

File: tests/cypress/integration/components/a-z-nav.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  });
+
+
+  // checking for class name - nhsuk-c-a-z-nav__list
+  it('Checking for class name - nhsuk-c-a-z-nav__list', function () {
+    cy
+      .get('#nhsuk-nav-a-z').children().should('have.class', 'nhsuk-nav-a-z__list')
+
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/a-z-nav.js]: Extra empty new line

File: tests/cypress/integration/components/a-z-nav.js [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  });
+
+  // checking for class name - nhsuk-c-a-z-nav__list-item-link
+  it('Checking for class name - nhsuk-nav-a-z__link', function () {
+    cy
+      .get('.nhsuk-nav-a-z__item').first().children().should('have.class', 'nhsuk-nav-a-z__link')
+
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/a-z-nav.js]: Extra empty new line

File: tests/cypress/integration/components/a-z-nav.js [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+  // checking for class name - nhsuk-c-a-z-nav__list
+  it('Checking for class name - nhsuk-c-a-z-nav__list', function () {
+    cy
+      .get('#nhsuk-nav-a-z').children().should('have.class', 'nhsuk-nav-a-z__list')
+
+  });
+
+  // checking for class name - nhsuk-c-a-z-nav__list-item-link
+  it('Checking for class name - nhsuk-nav-a-z__link', function () {
+    cy
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/a-z-nav.js]: Can this go on the same line cy.visit()?

File: tests/cypress/cypress.json (hunk 1 above)
Comment (suryapulakhandam) [reply tests/cypress/cypress.json]: updated

File: tests/cypress/integration/components/a-z-nav.js (hunk 2 above)
Comment (suryapulakhandam) [reply tests/cypress/integration/components/a-z-nav.js]: updated the description

File: tests/cypress/integration/components/a-z-nav.js (hunk 3 above)
Comment (suryapulakhandam) [reply tests/cypress/integration/components/a-z-nav.js]: removed the comments

File: tests/cypress/integration/components/a-z-nav.js (hunk 4 above)
Comment (suryapulakhandam) [reply tests/cypress/integration/components/a-z-nav.js]: removed the comments

File: tests/cypress/integration/components/a-z-nav.js (hunk 5 above)
Comment (suryapulakhandam) [reply tests/cypress/integration/components/a-z-nav.js]: removed the comments

File: tests/cypress/integration/components/a-z-nav.js (hunk 6 above)
Comment (suryapulakhandam) [reply tests/cypress/integration/components/a-z-nav.js]: removed the comments

File: tests/cypress/integration/components/a-z-nav.js (hunk 7 above)
Comment (suryapulakhandam) [reply tests/cypress/integration/components/a-z-nav.js]: removed the spacing

File: tests/cypress/integration/components/a-z-nav.js (hunk 8 above)
Comment (suryapulakhandam) [reply tests/cypress/integration/components/a-z-nav.js]: removed the extra line

File: tests/cypress/integration/components/a-z-nav.js (hunk 9 above)
Comment (suryapulakhandam) [reply tests/cypress/integration/components/a-z-nav.js]: removed the extra empty line

File: tests/cypress/integration/components/a-z-nav.js (hunk 10 above)
Comment (suryapulakhandam) [reply tests/cypress/integration/components/a-z-nav.js]: removed the empty line

File: tests/cypress/integration/components/a-z-nav.js (hunk 11 above)
Comment (suryapulakhandam) [reply tests/cypress/integration/components/a-z-nav.js]: removed the empty line

File: tests/cypress/integration/components/a-z-nav.js (hunk 12 above)
Comment (suryapulakhandam) [reply tests/cypress/integration/components/a-z-nav.js]: updated it into single line

**Comment by suryapulakhandam** [comment]
updated the test decription with the component name

File: package.json [hunk 13]
```diff
@@ -42,7 +46,8 @@
     "gulp-zip": "^4.2.0",
     "jquery": "^3.3.1",
     "nunjucks": "^3.1.3",
-    "sass-lint": "^1.12.1"
+    "sass-lint": "^1.12.1",
+    "start-server-and-test": "^1.7.9"
```
Comment (AdamChrimes) [inline package.json]: We can do the same thing without installing another dependency: `"test:cypress": "npm start && cypress",` We don't need to install "start-server-and-test": "^1.7.9" to do this. `npm uninstall start-server-and-test`

Comment (suryapulakhandam) [reply package.json]: updated with "test:cypress": "npm start & npm run cypress" and uninstalled start-server-and-test

File: tests/cypress/integration/components/a-z-nav.js [hunk 14]
```diff
@@ -0,0 +1,35 @@
+
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/a-z-nav.js]: This file has loads of random empty lines at the top.

File: tests/cypress/integration/components/a-z-nav.js [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+
+describe('A-Z Navigation Component', function () {
+
+
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/a-z-nav.js]: Random empty line

File: package.json [hunk 16]
```diff
@@ -1,13 +1,17 @@
 {
   "name": "nhsuk-frontend",
-  "version": "0.5.0",
+  "version": "0.1.6",
```
Comment (AdamChrimes) [inline package.json]: Make sure you get latest version from master, this shouldn't be changed to an earlier version.

File: tests/cypress/integration/components/button.js [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+  it('Navigating to Button Component', function () {
+    cy.visit(Cypress.config().baseUrl + '/button/index.html')
+  });
+
+
+  it('Checking for nhsuk-button class', function () {
+    cy.get('.nhsuk-button')
+  });
+
+
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/button.js]: Random empty line.

File: tests/cypress/integration/components/details.js [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+      .get('.nhsuk-details__text').should('have.attr','aria-hidden', 'true')
+  });
+
+  it('Checking expander click status', function () {
+    cy
+      .get('.nhsuk-details__summary').click()
+      .get('.nhsuk-details__summary').should('have.attr','aria-expanded', 'true')
+  });
+
+
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/details.js]: Random empty line - can you make sure all these files are consistent in terms of spacing and indenting?

File: tests/cypress/integration/components/warning-callout.js [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+
+  it('checking for warning callout class', function () {
+    cy.get('.nhsuk-warning-callout')
+  });
+
+  it('checking for warning callout label class', function () {
+    cy.get('.nhsuk-warning-callout').children().should('have.class', 'nhsuk-warning-callout__label')
+  });
+
+  it('Checking background colour of the label', function () {
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/warning-callout.js]: We don't need to check the background colour, BackstopJS can do that.

File: tests/cypress/integration/components/warning-callout.js [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+    cy.get('.nhsuk-warning-callout__label')
+      .should('have.css', 'background-color')
+      .and('eq', 'rgb(255, 235, 59)')
+  });
+  it('Checking background colour of the callout', function () {
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/warning-callout.js]: We don't need to check the background colour, BackstopJS can do that.

File: package.json [hunk 21]
```diff
@@ -43,7 +47,8 @@
     "gulp-zip": "^4.2.0",
     "jquery": "^3.3.1",
     "nunjucks": "^3.1.3",
-    "sass-lint": "^1.12.1"
+    "sass-lint": "^1.12.1",
+    "start-server-and-test": "^1.7.11"
```
Comment (AdamChrimes) [inline package.json]: Don't really like the idea of having another dependency that just starts the server. Can you investigate this further?

`"start-server": "npm start"` is "start": "gulp",

File: tests/cypress.json [hunk 22]
```diff
@@ -0,0 +1,7 @@
+{
+  "baseUrl": "http://localhost:3000/components/",
+  "liveUrl": "https://www.nhs.uk/",
```
Comment (AdamChrimes) [inline tests/cypress.json]: We aren't testing anything on the live website, do we need this configuration?

File: tests/cypress/integration/components/care-card-immediate.js [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+      .should('have.class', 'nhsuk-u-visually-hidden')
+  });
+
+  it('Checking care card title color to be red', function () {
+    cy.get('.nhsuk-care-card__heading-container')
+      .should('have.css', 'background-color')
+      .and('eq', 'rgb(218, 41, 28)')
+  });
+
+  it('Checking care card background color to be black', function () {
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/care-card-immediate.js]: Use BackstopJS for the visual stuff.

File: tests/cypress/integration/components/care-card-immediate.js [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+    cy.get('.nhsuk-care-card__content')
+      .should('have.css', 'background-color')
+      .and('eq', 'rgb(33, 43, 50)')
+  });
+
+  it('Checking font', function () {
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/care-card-immediate.js]: Use BackstopJS for the visual stuff, theres not much point in checking for the font.

File: tests/cypress/integration/components/warning-callout.js (hunk 20 above)
Comment (suryapulakhandam) [reply tests/cypress/integration/components/warning-callout.js]: Backstop checks whether the colour matches with the reference, but not the colour gradient - rgb(255, 235, 59)

File: tests/cypress/integration/components/warning-callout.js (hunk 19 above)
Comment (suryapulakhandam) [reply tests/cypress/integration/components/warning-callout.js]: Backstop checks whether the colour matches with the reference, but not the colour gradient - rgb(255, 235, 59)

**Comment by AdamChrimes** [comment]
@suryapulakhandam can you update your branch with master please

File: tests/cypress/integration/components/warning-callout.js (hunk 20 above)
Comment (AdamChrimes) [reply tests/cypress/integration/components/warning-callout.js]: Why do we need to check the gradient colour? Unless this is checking against a ratio for accessibility then it is pointless.

File: tests/cypress/integration/components/skip-link.js [hunk 25]
```diff
@@ -0,0 +1,12 @@
+describe('Skip link component tests', function () {
+  beforeEach(function () {
+    cy.viewport('macbook-15');
+  });
+
+  it('Navigating to skip link Component', function () {
+    cy.visit(Cypress.config().baseUrl + '/skip-link.html')
+  });
+
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/skip-link.js]: 2 empty lines here

File: tests/cypress/integration/components/button.js [hunk 26]
```diff
@@ -0,0 +1,22 @@
+describe('Button Component tests', function () {
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/button.js]: Can we check the consistency of the each of the components describes

Capitalisation of random letters and conventions

**Comment by AdamChrimes** [comment]
@suryapulakhandam are we doing any accessibility testing using any of the frameworks that Sab investigated? I see you have installed Axe as a dependency but it isn't used.

File: tests/cypress/integration/components/warning-callout.js (hunk 19 above)
Comment (AdamChrimes) [reply tests/cypress/integration/components/warning-callout.js]: But what is the purpose of checking the colour gradient? We aren't checking the colour contrast ratio for accessibility purpose so its a pointless test no? BackstopJS will pick up if the colour changes.

File: tests/cypress/integration/components/warning-callout.js [hunk 27]
```diff
@@ -0,0 +1,28 @@
...
+  it('Navigating to warning callout Component', function () {
+    cy.visit(Cypress.config().baseUrl + '/warning-callout.html')
+  });
+
+  it('Checking for warning callout class', function () {
+    cy.get('.nhsuk-warning-callout')
+  });
+
+  it('Checking for warning callout label class', function () {
+    cy.get('.nhsuk-warning-callout').children().should('have.class', 'nhsuk-warning-callout__label')
```
Comment (AdamChrimes) [inline tests/cypress/integration/components/warning-callout.js]: Can this check for content aswell? To make sure that there is content within the label such as 'Important' and not just the class name exists. Because in theory you are just checking the class name exists, not that there is a heading.
